grpc/grpc_server.cpp

```diff
@@ -0,0 +1,83 @@
+#include "requests_common.hpp"
+
+#include "read_request.hpp"
+#include "write_request.hpp"
+
+namespace fb = flatbuffers;
+
+namespace grpc_dnet {
+
+struct server {
+public:
+	server(dnet_node *node)
+		: node_(node)
+	{}
+
+	~server() {
+		server_->Shutdown();
+		cq_->Shutdown();
+		thread_->join();
+	}
+
+	void run() {
+		std::string server_address("0.0.0.0:2025");
+
+		grpc::ServerBuilder builder;
+		builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
+		builder.RegisterService(&service_);
+
+		cq_ = builder.AddCompletionQueue();
+
+		server_ = builder.BuildAndStart();
+
+		DNET_LOG_INFO(node_, "GRPC: server listening on {}", server_address);
+
+		thread_ = std::make_unique<std::thread>(std::bind(&server::completion_thread, this));
+	}
+
+	void completion_thread() {
+		service_provider_info provider;
+		provider.node = node_;
+		provider.service = &service_;
+		provider.cq = cq_.get();
+
+		new read_request_manager(&provider,
+		                         [](std::unique_ptr<ell::n2::read_request>){ /*TODO: implement*/ });
+		new write_request_manager(&provider,
+		                          [](std::unique_ptr<ell::n2::write_request>){ /*TODO: implement*/ });
+
+		void *tag;
+		bool ok;
+		while (cq_->Next(&tag, &ok)) {
+			static_cast<request_manager_base*>(tag)->process_completion(ok);
+		}
+	}
+
+private:
+	dnet_node *node_;
+	std::unique_ptr<grpc::ServerCompletionQueue> cq_;
+	fb_grpc_dnet::Elliptics::AsyncService service_;
+	std::unique_ptr<grpc::Server> server_;
+	std::unique_ptr<std::thread> thread_;
+};
+
+} // namespace grpc_dnet
+
+// Out of namespace: C compatible API
+
+struct dnet_grpc_server {
+	grpc_dnet::server server;
```

Comment:
Why is it wrapped into `dnet_grpc_server`?

Reply:
To have grpc server class situated out of namespace (for C users), because we cannot put in C forward declaration `struct grpc_dnet::server;`

Reply:
Why do you need a class situated in namespace?

Reply:
To avoid many occurences of `ioremap::elliptics` in C++ code (header/implementation). You told recently, that it's default project namespace. So I think we must to aim to embrace as much C++ code as we can in this namespace.